// @ethersphere/bee-js/dist/mjs/modules/debug/connectivity.js
import { http } from "../../utils/http.js";
export async function getNodeAddresses(requestOptions) {
  const response = await http(requestOptions, {
    url: 'addresses',
    responseType: 'json'
  });
  return response.data;
}
export async function getPeers(requestOptions) {
  const response = await http(requestOptions, {
    url: 'peers',
    responseType: 'json'
  });
  return response.data.peers;
}
export async function getBlocklist(requestOptions) {
  const response = await http(requestOptions, {
    url: 'blocklist',
    responseType: 'json'
  });
  return response.data.peers;
}
export async function removePeer(requestOptions, peer) {
  const response = await http(requestOptions, {
    url: `peers/${peer}`,
    responseType: 'json',
    method: 'DELETE'
  });
  return response.data;
}
export async function getTopology(requestOptions) {
  const response = await http(requestOptions, {
    url: `topology`,
    responseType: 'json'
  });
  return response.data;
}
export async function pingPeer(requestOptions, peer) {
  const response = await http(requestOptions, {
    url: `pingpong/${peer}`,
    responseType: 'json',
    method: 'POST'
  });
  return response.data;
}
