// @ethersphere/bee-js/dist/mjs/modules/debug/stake.js
import { http } from "../../utils/http.js";
const STAKE_ENDPOINT = 'stake';
const REDISTRIBUTION_ENDPOINT = 'redistributionstate';
/**
 * Gets the staked amount
 *
 * @param requestOptions Options for making requests
 */
export async function getStake(requestOptions) {
  const response = await http(requestOptions, {
    method: 'get',
    responseType: 'json',
    url: `${STAKE_ENDPOINT}`
  });
  return response.data.stakedAmount.toString();
}
/**
 * Stake given amount of tokens.
 *
 * @param requestOptions Options for making requests
 * @param amount
 * @param options
 */
export async function stake(requestOptions, amount, options) {
  const headers = {};
  if (options?.gasPrice) {
    headers['gas-price'] = options.gasPrice.toString();
  }
  if (options?.gasLimit) {
    headers['gas-limit'] = options.gasLimit.toString();
  }
  await http(requestOptions, {
    method: 'post',
    responseType: 'json',
    url: `${STAKE_ENDPOINT}/${amount}`,
    headers
  });
}
/**
 * Get current status of node in redistribution game
 *
 * @param requestOptions Options for making requests
 */
export async function getRedistributionState(requestOptions) {
  const response = await http(requestOptions, {
    method: 'get',
    responseType: 'json',
    url: REDISTRIBUTION_ENDPOINT
  });
  return response.data;
}
