// @ethersphere/bee-js/dist/mjs/modules/debug/transactions.js
import { http } from "../../utils/http.js";
const transactionsEndpoint = 'transactions';
/**
 * Get list of all pending transactions
 *
 * @param requestOptions Options for making requests
 */
export async function getAllTransactions(requestOptions) {
  const response = await http(requestOptions, {
    url: transactionsEndpoint,
    responseType: 'json'
  });
  return response.data.pendingTransactions;
}
/**
 * Get information for specific pending transactions
 *
 * @param requestOptions Options for making requests
 * @param transactionHash Hash of the transaction
 */
export async function getTransaction(requestOptions, transactionHash) {
  const response = await http(requestOptions, {
    url: `${transactionsEndpoint}/${transactionHash}`,
    responseType: 'json'
  });
  return response.data;
}
/**
 * Rebroadcast existing transaction
 *
 * @param requestOptions Options for making requests
 * @param transactionHash Hash of the transaction
 */
export async function rebroadcastTransaction(requestOptions, transactionHash) {
  const response = await http(requestOptions, {
    method: 'post',
    url: `${transactionsEndpoint}/${transactionHash}`,
    responseType: 'json'
  });
  return response.data.transactionHash;
}
/**
 * Cancel existing transaction
 *
 * @param requestOptions Options for making requests
 * @param transactionHash Hash of the transaction
 * @param gasPrice Optional gas price
 */
export async function cancelTransaction(requestOptions, transactionHash, gasPrice) {
  const headers = {};
  if (gasPrice) {
    headers['gas-price'] = gasPrice;
  }
  const response = await http(requestOptions, {
    method: 'delete',
    headers,
    url: `${transactionsEndpoint}/${transactionHash}`,
    responseType: 'json'
  });
  return response.data.transactionHash;
}
