// @ethersphere/bee-js/dist/mjs/modules/soc.js
import { extractUploadHeaders } from "../utils/headers.js";
import { http } from "../utils/http.js";
const socEndpoint = 'soc';
/**
 * Upload single owner chunk (SOC) to a Bee node
 *
 * @param requestOptions  Options for making requests
 * @param owner           Owner's ethereum address in hex
 * @param identifier      Arbitrary identifier in hex
 * @param signature       Signature in hex
 * @param data            Content addressed chunk data to be uploaded
 * @param postageBatchId  Postage BatchId that will be assigned to uploaded data
 * @param options         Additional options like tag, encryption, pinning
 */
export async function upload(requestOptions, owner, identifier, signature, data, postageBatchId, options) {
  const response = await http(requestOptions, {
    method: 'post',
    url: `${socEndpoint}/${owner}/${identifier}`,
    data,
    headers: {
      'content-type': 'application/octet-stream',
      ...extractUploadHeaders(postageBatchId, options)
    },
    responseType: 'json',
    params: {
      sig: signature
    }
  });
  return response.data.reference;
}
