// @ethersphere/bee-js/dist/mjs/modules/tag.js
import { http } from "../utils/http.js";
const endpoint = 'tags';
/**
 * Create new tag on the Bee node
 *
 * @param url Bee tag URL
 */
export async function createTag(requestOptions) {
  const response = await http(requestOptions, {
    method: 'post',
    url: endpoint,
    responseType: 'json'
  });
  return response.data;
}
/**
 * Retrieve tag information from Bee node
 *
 * @param url Bee tag URL
 * @param uid UID of tag to be retrieved
 */
export async function retrieveTag(requestOptions, uid) {
  const response = await http(requestOptions, {
    url: `${endpoint}/${uid}`,
    responseType: 'json'
  });
  return response.data;
}
/**
 * Get limited listing of all tags.
 *
 * @param url
 * @param offset
 * @param limit
 */
export async function getAllTags(requestOptions, offset, limit) {
  const response = await http(requestOptions, {
    url: endpoint,
    params: {
      offset,
      limit
    },
    responseType: 'json'
  });
  return response.data.tags;
}
/**
 * Removes tag from the Bee node.
 * @param url
 * @param uid
 */
export async function deleteTag(requestOptions, uid) {
  await http(requestOptions, {
    method: 'delete',
    url: `${endpoint}/${uid}`
  });
}
/**
 * Updates tag
 * @param url
 * @param uid
 * @param reference
 */
export async function updateTag(requestOptions, uid, reference) {
  await http(requestOptions, {
    method: 'patch',
    url: `${endpoint}/${uid}`,
    data: {
      reference
    }
  });
}
