// @ethersphere/bee-js/dist/mjs/utils/http.js
import axios from 'axios';
import { Objects, Strings } from 'cafe-utility';
import { BeeResponseError } from "../index.js";
const {
  AxiosError
} = axios;
export const DEFAULT_HTTP_CONFIG = {
  headers: {
    accept: 'application/json, text/plain, */*'
  },
  maxBodyLength: Infinity,
  maxContentLength: Infinity
};
/**
 * Main function to make HTTP requests.
 * @param options User defined settings
 * @param config Internal settings and/or Bee settings
 */
export async function http(options, config) {
  try {
    const requestConfig = Objects.deepMerge3(DEFAULT_HTTP_CONFIG, config, options);
    maybeRunOnRequestHook(options, requestConfig);
    const response = await axios(requestConfig);
    // TODO: https://github.com/axios/axios/pull/6253
    return response;
  } catch (e) {
    if (e instanceof AxiosError) {
      throw new BeeResponseError(e.message, e.status, e.code);
    }
    throw e;
  }
}
function maybeRunOnRequestHook(options, requestConfig) {
  if (options.onRequest) {
    options.onRequest({
      method: requestConfig.method || 'GET',
      url: Strings.joinUrl(requestConfig.baseURL, requestConfig.url),
      headers: {
        ...requestConfig.headers
      },
      params: requestConfig.params
    });
  }
}
